### PR TITLE
Added a cloud renderer that uploads geometry to the GPU.

### PR DESCRIPTION
**Pull request by Zaggy1024**
#3188 updated to 1.11.x. All comments on that PR have been addressed.

Code in CloudRenderer is cleaned up a bit to rebuild the VBOs/display lists as little as possible, in a client tick handler rather than the renderer.

To avoid adding unnecessary profiling lines EntityRenderer, ProfilerEnableEvent was added to temporarily force on the profiler to time the rendering of the clouds. This can also be reused for other profiling test mods in the future, if any such tests are needed.

### CODE REVIEW

**Comment by mezz** [comment]
Please update this PR for the latest version of Forge, the MCP mappings have been updated.

**Comment by mezz** [comment]
Also `Math.floorMod` is not available in Java 6, so you should copy that method somewhere.

**Comment by liach** [comment]
@mezz But I don't see `Math.floorMod` in this pull request.

File: src/main/java/net/minecraftforge/client/CloudRenderer.java [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+        int offV = fullCoord(z, scale);
+
+        GlStateManager.pushMatrix();
+
+        // Translate by the remainder after the UV offset.
+        GlStateManager.translate((offU * scale) - x, y, (offV * scale) - z);
+
+        // Modulo to prevent texture samples becoming inaccurate at extreme offsets.
+        offU = Math.floorMod(offU, texW);
+        offV = Math.floorMod(offV, texH);
```
Comment (mezz) [inline src/main/java/net/minecraftforge/client/CloudRenderer.java]: Right here @liach

**Comment by Zaggy1024** [comment]
Fixed both issues. :)

**Comment by bs2609** [comment]
Is the change from `floorMod` to `%` ok?

**Comment by Zaggy1024** [comment]
Yep. Not sure why I went with that function in the first place to be honest.

**Comment by bs2609** [comment]
As long as negative values are fine, it's not a problem. If you do actually need it, Guava has `IntMath.mod()`

**Comment by mezz** [comment]
I tested and saw a couple small issues:

1. You should make the test mod disabled by default, just add a static field `ENABLED = false` and put the test mod's event subscription line behind `if (ENABLED)`

2. It's very improper to have something that isn't a mod subscribe itself to the event bus.
You should find another way, or just have forge initialize the cloud renderer at a more appropriate time so that FML can determine which "mod" (forge) is registering your cloud renderer.

```
[00:19:18] [Client thread/ERROR] [FML]: Unable to determine registrant mod for INSTANCE. This is a critical error and should be impossible
java.lang.Throwable
	at net.minecraftforge.fml.common.eventhandler.EventBus.register(EventBus.java:76) [EventBus.class:?]
	at net.minecraftforge.client.CloudRenderer.<init>(CloudRenderer.java:82) [CloudRenderer.class:?]
	at net.minecraftforge.client.CloudRenderer.<clinit>(CloudRenderer.java:50) [CloudRenderer.class:?]
	at net.minecraft.client.renderer.RenderGlobal.renderClouds(RenderGlobal.java:1468) [RenderGlobal.class:?]
	at net.minecraft.client.renderer.EntityRenderer.renderCloudsCheck(EntityRenderer.java:1439) [EntityRenderer.class:?]
	at net.minecraft.client.renderer.EntityRenderer.renderWorldPass(EntityRenderer.java:1295) [EntityRenderer.class:?]
	at net.minecraft.client.renderer.EntityRenderer.renderWorld(EntityRenderer.java:1246) [EntityRenderer.class:?]
	at net.minecraft.client.renderer.EntityRenderer.updateCameraAndRender(EntityRenderer.java:1053) [EntityRenderer.class:?]
	at net.minecraft.client.Minecraft.runGameLoop(Minecraft.java:1075) [Minecraft.class:?]
	at net.minecraft.client.Minecraft.run(Minecraft.java:370) [Minecraft.class:?]
	at net.minecraft.client.main.Main.main(Main.java:118) [Main.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_102]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_102]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_102]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_102]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_102]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_102]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_102]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_102]
	at net.minecraftforge.gradle.GradleStartCommon.launch(GradleStartCommon.java:97) [Forge_main/:?]
	at GradleStart.main(GradleStart.java:26) [Forge_main/:?]
```

Test results look promising:
```
[00:19:40] [Client thread/INFO] [FML]: Forge clouds took 0.103533596ms per frame average
[00:19:50] [Client thread/INFO] [FML]: Vanilla clouds took 1.1159655ms per frame average
[00:20:00] [Client thread/INFO] [FML]: Forge clouds took 0.06136547ms per frame average
[00:20:10] [Client thread/INFO] [FML]: Vanilla clouds took 1.4785231ms per frame average
[00:20:10] [Client thread/INFO] [FML]: Switched back to Forge clouds
```

**Comment by Zaggy1024** [comment]
I figured for the profiling there's really no harm in letting it run by default, since the test mods will only be in the Forge contrib workspaces (if I recall correctly). With it on by default, hopefully any changes in its performance relative to vanilla will be noticed more readily, as well. In case that makes a difference.

Where would be a good place to call the initialization from? ForgeModContainer, but behind a side check?

**Comment by williewillus** [comment]
all the test mods should be off by default, it's just a convention. so contributors aren't confused why a spawn egg does nothing because the PIE test mod cancels its use, for example

**Comment by Zaggy1024** [comment]
Ah, I didn't realize others were disabled by default... however, this one isn't intrusive and it doesn't break anything, should it be the same?

**Comment by mezz** [comment]
Yeah it should be off by default. The test mods are mainly for debugging when something is wrong.

**Comment by Zaggy1024** [comment]
Okay, now the only question I have left is where would be the best place to initialize the renderer?

**Comment by mezz** [comment]
It looks like `ForgeModContainer.preInit` would be a good spot, several other Forge things are registered there.

**Comment by Zaggy1024** [comment]
Both issues solved, although I had to use postInit to avoid an NPE.

**Comment by Zaggy1024** [comment]
Looks like calling the init from ForgeModContainer is causing a class loading error on the server, I'll have to figure out how to solve that..

**Comment by mezz** [comment]
To solve that, you should not have `public CloudRenderer()` register itself. Make `ForgeModContainer` register it at appropriate times.
If you register your reload listener too early, it will get called during startup (possibly multiple times).

**Comment by mezz** [comment]
Also `IResourceManagerReloadListener` is client side only, so anything that references `CloudRenderer` needs to be client-side.

**Comment by mezz** [comment]
It may be cleaner if you register a separate `IResourceManagerReloadListener` anonymous class that calls `reloadTextures()` on the cloud render instance.

**Comment by Zaggy1024** [comment]
Disregard the previous comment, it appears that the error was caused by CustomSpriteTest. The current revision functions perfectly on the client and the server, according to my testing.

**Comment by Zaggy1024** [comment]
Is there anything else to change, or is it ready?

**Comment by mezz** [comment]
Does not compile, you have a field `instance` and try to get `INSTANCE` in `RenderGlobal`.
Once you've fixed that, I've tested it and everything looks good.

**Comment by Zaggy1024** [comment]
Ah, forgot to genPatches... Fixed.

**Comment by mezz** [comment]
It's your lucky day, 1.11.2 came out! Can you update your patches once more?

**Pull request closed without merging**